cpp/src/arrow/acero/unmaterialized_table.h

```diff
@@ -0,0 +1,273 @@
+
+  Result<std::optional<std::shared_ptr<RecordBatch>>> Materialize() {
+    // Don't build empty batches
+    if (Empty()) {
+      return std::nullopt;
+    }
+    DCHECK_LE(Size(), (uint64_t)std::numeric_limits<int64_t>::max());
+    std::vector<std::shared_ptr<arrow::Array>> arrays(schema->num_fields());
+
+#define MATERIALIZE_CASE(id)                                                          \
+  case arrow::Type::id: {                                                             \
+    using T = typename arrow::TypeIdTraits<arrow::Type::id>::Type;                    \
+    ARROW_ASSIGN_OR_RAISE(arrays.at(i_col), materializeColumn<T>(field_type, i_col)); \
+    break;                                                                            \
+  }
+
+    // Build the arrays column-by-column from the rows
+    for (int i_col = 0; i_col < schema->num_fields(); ++i_col) {
+      const std::shared_ptr<arrow::Field>& field = schema->field(i_col);
+      const auto& field_type = field->type();
+
+      switch (field_type->id()) {
+        MATERIALIZE_CASE(BOOL)
+        MATERIALIZE_CASE(INT8)
+        MATERIALIZE_CASE(INT16)
+        MATERIALIZE_CASE(INT32)
+        MATERIALIZE_CASE(INT64)
+        MATERIALIZE_CASE(UINT8)
+        MATERIALIZE_CASE(UINT16)
+        MATERIALIZE_CASE(UINT32)
+        MATERIALIZE_CASE(UINT64)
+        MATERIALIZE_CASE(FLOAT)
+        MATERIALIZE_CASE(DOUBLE)
+        MATERIALIZE_CASE(DATE32)
+        MATERIALIZE_CASE(DATE64)
+        MATERIALIZE_CASE(TIME32)
+        MATERIALIZE_CASE(TIME64)
+        MATERIALIZE_CASE(TIMESTAMP)
+        MATERIALIZE_CASE(STRING)
+        MATERIALIZE_CASE(LARGE_STRING)
+        MATERIALIZE_CASE(BINARY)
+        MATERIALIZE_CASE(LARGE_BINARY)
+        default:
+          return arrow::Status::Invalid("Unsupported data type ",
+                                        field->type()->ToString(), " for field ",
+                                        field->name());
+      }
+    }
+
+#undef MATERIALIZE_CASE
+
+    std::shared_ptr<arrow::RecordBatch> r =
+        arrow::RecordBatch::Make(schema, (int64_t)num_rows, arrays);
+    return r;
+  }
+
+ private:
+  struct UnmaterializedSlice {
+    CompositeEntry components[MAX_COMPOSITE_TABLES];
+    size_t num_components;
+
+    inline int64_t Size() const {
+      if (num_components == 0) {
+        return 0;
+      }
+      return components[0].end - components[0].start;
+    }
+  };
+
+  // Mapping from an output column ID to a source table ID and column ID
+  std::shared_ptr<arrow::Schema> schema;
+  size_t num_composite_tables;
+  std::unordered_map<int, std::pair<int, int>> output_col_to_src;
+
+  arrow::MemoryPool* pool;
+
+  /// A map from address of a record batch to the record batch. Used to
+  /// maintain the lifetime of the record batch in case it goes out of scope
+  /// by the main exec node thread
+  std::unordered_map<uintptr_t, std::shared_ptr<arrow::RecordBatch>> ptr2Ref = {};
```

Comment:
When do entries get removed from this map?

Reply:
Only when the `UnmaterializedTable` is destructed. There's perhaps an argument that we can clear out when we materialize into a real table since we don't need the references after we create an output record batch, but I was wary of this approach (what if one day we want to materialize twice?)

Reply:
I'm probably missing something but does this mean it will accumulate the entire input into memory?  One advantage of a sorted merge node is that it can process a very large stream of data, data that is larger than the memory on the device.  If we are keeping references to all the record batches then will we still be able to do that?

Reply:
Good Q! This is actually a generalization/replacement of the `CompositeReferenceTable` in asofjoin today: https://github.com/apache/arrow/blob/main/cpp/src/arrow/acero/asof_join_node.cc#L1282

The `CompositeReferenceTable` also stores a `ptr2ref` object, which you're right it will keep a set of record batches alive for its lifetime. However, these objects are only constructed ephemerally as locals while we output a batch of data.

In the case of the asofjoin node, we create a `CompositeTableBuilder` local, which wraps this object when we have buffered enough data to create a new output batch.

In the case of sorted_merge, it's similar in that we wait until we have some data from all the inputs, and then ephemerally construct an `UnmaterializedCompositeTable` and destruct it as soon as it's materialized.

What would be bad is if we kept the ptr2ref mapping alive longer than needed, but since we construct new instances of this class for each batch, record batches go out of scope appropriately.

Reply:
Ah, thank you.  The piece I was missing was that we destruct the `UnmaterializedCompositeTable` after materialization.